Lib/ast.py

```diff
@@ -1053,15 +1054,77 @@ def visit_AsyncWith(self, node):
         with self.block(extra=self.get_type_comment(node)):
             self.traverse(node.body)
 
+    def _str_literal_helper(
+        self, value, quote_types=("'", '"', '"""', "'''"), escape=""
+    ):
+        """Helper for writing string literals, minimising escapes.
+        Returns (possible quote types, string literal to write).
+        """
+        # Escape characters we've been told to escape, backslashes, and
+        # non-printable characters other than \n and \t.
+        escape = {*escape, '\\'}
+        val = "".join(
+            c.encode('unicode_escape').decode('ascii')
+            if c in escape or (not c.isprintable() and c not in '\n\t') else c
+            for c in value
+        )
+        qts = [quote for quote in quote_types if quote not in val]
+        if "\n" in val:
+            qts = [quote for quote in qts if quote in ('"""', "'''")]
+        if not qts:
+            # If there aren't any possible quote_types, fallback to using repr
+            # on the original value. Try to use a quote_type from quote_types.
+            value = repr(value)
+            quote_type = next((q for q in quote_types if value[0] in q), value[0])
+            return value[1:-1], [quote_type]
```

Comment:
Would be nice to move this function's body to an inner function called `fallback` and return it there, so we can replace the `assert len(qts[0]) == 3` check with this
```
        def fallback(original):
            quote, *escaped_string_parts, quote = repr(original)
            for possible_quote in quote_types:
                if quote in possible_quote:
                    break
            else:
                possible_quote = quote
            return "".join(escaped_string_parts), possible_quote
```

Reply:
(Resolving this, since I replied on the other related comment https://github.com/python/cpython/pull/19612#discussion_r436394009)